07/hangman/pictures.py

```diff
@@ -0,0 +1,37 @@
+
+def picture(fail_try):
```

Comment:
Idea se soubory, ze kterých načítat obrázky a pak je renderovat, je dobrá. Ale implementace by mohla být lepší 😃.

1. Funkce `draw_picture(...)` nic nevrací _(technicky vzato implicitně vrací `None`)_, takže `return draw_picture(...)` bude taky dycky vracet `None`, čili z téhle funkce `picture(...)` se taky dycky bude vracet `None`.
	- No a jelikož jediné místo, kde `picture(...)` voláš, je zde: https://github.com/irenaskl/pyladies/pull/5/files#diff-7b14f189559dc5c917ec1df9458050ac92325d46340645ae7d863f86753ad4c3R19 ... kde s návratovou hodnotou _(která je stejně dycky `None`)_ stejně nic neděláš, tak tyhle `return`y v `return draw_picture(...)` jsou úplně zbytečně.
		- Úplně v pohodě můžeš prostě volat `draw_picture(...)` a nemít před tím `return`.
2. Tahle logika mapování z hodnot `fail_try` _(od 0 až do 9)_ na názvy souborů _(`picture_zero.txt` až `picture_nine,txt`)_ jsou přece ***ideální kandidát*** na to, použít ***slovník*** _(v jiných okruzích se slovníkům dokonce říká přímo "mapy")_.
	- Mít tady tenhle X-násobný (EL-)IF je přece strašně ošklivé - a neříkej mi, že tě to taky nenapadlo, že by to šlo určitě udělat nějak hezčí - ***a obecnější*** 😉 
	- Buď jsi mohla prostě natvrdo zadefinovaný slovník mapující `{0: 'picture_zero.txt', 1: 'picture_one.txt, ...'}`¨, který bys použila místo tohohle hrozného repetetivního IFu.
	- Nebo přece šlo ty soubory přímo tak nazvat _(jakože přece klidně to místo `picture_one.txt` mohlo být `picture_0.txt`, žejo 😄)_ - a místo nějakého natvrdo mapování jsi mohla mít něco jako:
	```python
	def picture(index):
		picture_filename = f"picture_{index}.txt"
		...
	```
3. Soubory přes `open(...)` sice otevíráš, ale ***nikdy nezavíráš*** 😨 A soubory bys zavírat měla.
	- Přeči si třeba tady: https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files či https://realpython.com/why-close-file-python/
	- Funkce `open(...)` vrací _file handler_, čili když napíšeš `game_picture = open(...)`, tak v proměnné `game_picture` je uložený _file handler_ objekt. Přes ten jde číst _(což děláš - tím, že přes něj iteruješ)_, ale taky se přes něj dá soubor přímo zavřít - v tvém případě tedy např. přes `game_picture.close()` - což neděláš 😉 Ale měla bys.
		- Druhou možností je použít _context manager_ _(což je taková vychytávka v Pythonu a už jste to viděli i na hodině - jedná se o konstrukt `with <expr>: ...`)_.
			- Víc o context managerech třeba zde: https://realpython.com/python-with-statement/
			- V tvém případě se by se dalo využít toho, že **file handler se přímo dá použít jako context manger**, který se v takovém případě sám umí ve finále zavřít:
				```python
				with open('picture1.txt') as fh:
					for line in fh:
						print(line, end='')

				# A v tomhle místě (venku z "with") už se 'automaticky'
				# zavolalo "fh.close()"
				```

Ve finále bych to já udělal třeba nějak takhle:

```python
def picture(index):
	filename = f"picture.{index}.txt"
	try:
		with open(filename, 'r') as fh:
			for line in fh:
				print(line, end='')

	except OSError:
		# Protože jsem si přečetl dokumentaci funkce open() a vím, že může
		# vyhodit OSError (viz https://docs.python.org/3/library/functions.html#open
		# a https://docs.python.org/3/library/exceptions.html#exception-hierarchy)
		print(f"Couldn't read picture from '{filename}")

```

... s tím, že soubory s obrázky bych měl pojmenované tak, že se mi s nimi v kódu bude dobře pracovat _(tedy mít v nich radši přímo čísla, než zbytečně slova)_ a také ***počítám s možnou výjimkou*** vyhozenou funkcí `open()` - a taky ji mám správně ošetřenou _(tak, že jsem si já řekl, že pokud by se NÁHODOU nějak stalo, že by se měl vykreslit neexistující obrázek, tak vypíšu nějakou chybovou hlášku a spokojím se s tím, že se obrázek prostě nevykreslí. Ale taky mi takhle nespadne komplet celý program kvůli chybějícímu obrázku, což je fajn.)_.